### DISCUSSION TITLE
workflow for iterating over formula development? any best practices? 🤔

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#3568] [Tap maintenance and Homebrew development]
been trying my best to keep the homebrew-freecad tap updated. it's far from perfect, but i feel confident in saying that it is in a better place now than when i started maintaining the tap. that said, i do have some gripes about formula development.

1. when updating a formula, and trying out new things, the development process seems clunky.

a very crude way of iterating / updating a formula is to have a text editor open and making changes to the formula file, and then running `brew install [NAME_OF_FORMULA]` this becomes a very long winded process especially if one needs to build the software and then run the software afterwards to to test their changes.

for my use case, (presently) there is a bug with the current batch of freecad bottles i published yesterday where freecad will crash when saving a document that contains a sketch. this is an issue i have experienced in the past. and my gut reaction is that i believe i need to add the path of llvm to the list of _cmake prefix path_ during the _configure_ setup process of the build. however, if this doesn't fix my issue, i'll have waited ~ over an hour for the build to complete only for brew to wipe out my _build_ directory and then any subsequent changes i make to such formula file i have to start the entire process over again. this is far from ideal.

i'd say this almost begs the question, should homebrew support saving / restoring of _build_ directories to cut down on _rebuild_ times. (forgive me for lay terms there). i know the _install_ subcommand to _brew_ supports a `--keep-tmp` flag but i don't think brew in its current form supports restoring of that _tmp_ directory 🧐.

i guess the best (current)  way for iterating / testing of formula development is to run the `brew install` commands with the `--debug` flag and then manually running the `cmake` `make` `make install` process over & over until one achieves the desired results?

i'm open to any other suggestions 🤝 and hacks if anyone is using / doing something different to speed up the process of formula development.

**Comment by carlocab** [comment]
> i need to add the path of llvm to the list of _cmake prefix path_ during the _configure_ setup process of the build.

`brew` already does this. You can check by doing something like
```ruby
ohai "CMAKE_PREFIX_PATH = #{ENV["CMAKE_PREFIX_PATH"]}"
```
at the start of the `install` method.

> i guess the best (current) way for iterating / testing of formula development is to run the `brew install` commands with the `--debug` flag and then manually running the `cmake` `make` `make install` process over & over until one achieves the desired results?

Yes, that sounds right. `brew` is optimised for installing things, and not really for developing the `install` method.

It might be easier to work out your issues outside `brew` and then translate your build process into the formula.

**Reply by ipatch** [reply]
> It might be easier to work out your issues outside `brew` and then translate your build process into the formula.

from my experience that process does not work well going either way. if you have a working _cmake dance_ outside of _brew install_ it doesn't translate to a 1:1 `brew install`. certain things are lost or don't carry over. especially since the `brew install` is unaware of the `brew` command itself. and if certain formula are linked or unlinked then that can lend to more issues.

i'll mess around with the `brew install --debug` and see what i can come up with.
